### PR TITLE
Sync HPackDecoder with ASP.NET Core changes

### PR DESCRIPTION
**Pull request by JamesNK**
Sync with HPackDecoder changes in https://github.com/dotnet/aspnetcore/pull/24730

~Define `HPACK_SUPPORT_ONSTATICINDEXEDHEADER` to use new behavior. runtime doesn't define the variable so it is unchanged.~

~In .NET 6 runtime can optimize around the static index and the compiler variable and older code path can be removed.~

### CODE REVIEW

**Comment by geoffkizer** [comment]
As you mentioned in the ASP.NET PR, it seems like just fixing the HTTP2 code here to deal with these callbacks in a simplistic way is pretty trivial. Can we just add that code and avoid the #ifdefs?

**Comment by JamesNK** [comment]
@geoffkizer Updated

File: src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackDecoder.cs [hunk 1]
```diff
@@ -619,9 +648,7 @@ private static bool IsHuffmanEncoded(byte b)
         {
```
Comment (geoffkizer) [inline src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackDecoder.cs:648]: Nit: this function is now really GetDynamicHeader, right? Can we rename?

**Comment by geoffkizer** [comment]
@jamesnk Please make sure outerloop tests are run before merging

**Review by geoffkizer — APPROVED** [review]
One naming nit above, otherwise LGTM

**Pull request merged**